applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/differentiation_utility.h

```diff
@@ -0,0 +1,88 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Martin Fusseder, https://github.com/MFusseder
+//
+
+#if !defined(KRATOS_DIFFERENTIATION_UTILITY_H_INCLUDED )
+#define  KRATOS_DIFFERENTIATION_UTILITY_H_INCLUDED
+
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "includes/element.h"
+#include "includes/condition.h"
+#include "structural_mechanics_application_variables.h"
+
+namespace Kratos
+{
+
+/** \brief DifferentiationUtility
+ *
+ * This class calculates comnputes the derivatives of different element and condition entities
+ * (e.g. RHS, LHS, mass-matrix, ...) with respect to a design variable (e.g. nodal-coordinate, property).
+ */
+class DifferentiationUtility
+{
+public:
+
+    typedef std::size_t IndexType;
+    typedef std::size_t SizeType;
+    typedef VariableComponent<VectorComponentAdaptor<array_1d<double, 3>>> array_1d_component_type;
+
+    static void CalculateRigthHandSideDerivative(Element& rElement,
+                                                const Variable<double>& rDesignVariable,
+                                                const double& rPertubationSize,
+                                                Matrix& rOutput,
+                                                const ProcessInfo& rCurrentProcessInfo);
+
+    static void CalculateRigthHandSideDerivative(Element& rElement,
+                                                const Variable<array_1d<double,3>>& rDesignVariable,
+                                                const double& rPertubationSize,
+                                                Matrix& rOutput,
+                                                const ProcessInfo& rCurrentProcessInfo);
+
+    static void CalculateRigthHandSideDerivative(Element& rElement,
```

Comment:
Do we need both versions? `Variable<array_1d<double,3>>``and `array_1d_component_type`?

Reply:
I think in the future ```array_1d_component_type``` will be enough. The ```Variable<array_1d<double,3>>``` is equivalent to what currently the adjoint elements delivers. But the ```CalclateSensitivityMatrix" function can be easily redesigned that it can be used with ```array_1d_component_type``` shape-variables.

Reply:
we can also have keep the `Variable<array_1d<double,3>>` variant and then call the `array_1d_component_type` from within. Just i don't like to have the code duplication we have now.

Reply:
I agree with you. I will modify the ```CalculateSensitivityMatrix``` function of the adjoint base element.